### PR TITLE
Issue#1152

### PR DESCRIPTION
**Pull request by baala108**
Implemented Recursion limit counter and added test case.

### CODE REVIEW

**Comment by TimLethbridge** [comment]
@bala-v09 there is a failure in ParserConstraintExpressionsTest. You should see this when you do a full build. Please fix.

File: cruise.umple/src/Umple_Code.ump [hunk 1]
```diff
@@ -1880,8 +1880,15 @@ class UmpleClass
     return wasSet;
   }
 
+
+  internal const Integer CounterLimit = 25; //Arbitary Limit
+  Integer propagateCounter = CounterLimit;
```
Comment (TimLethbridge) [inline cruise.umple/src/Umple_Code.ump:1885]: You are using an instance variable for the propagateCounter. A better solution would have been to simply add an argument. The initial call sets it to 25; recursive calls reduce by1.
You are using an instance variable for the propagateCounter. A better solution would have been to simply add an argument. The initial call sets it to 25; recursive calls reduce by 1.

File: cruise.umple/test/cruise/umple/compiler/UmpleParserTest.java [hunk 2]
```diff
@@ -444,7 +444,21 @@ public void immutableClass()
     Assert.assertEquals(true,model.getUmpleClass("Student").isImmutable());
     Assert.assertEquals(false,model.getUmpleClass("Mentor").isImmutable());
   }
-
+  
+  
+  //Issue 1152
+  @Test
+  public void propagateImmutabilityToAllRelationships(){
+    assertSimpleParse("022_propagateImmutabilityToAllRelationships.ump");
```
Comment (TimLethbridge) [inline cruise.umple/test/cruise/umple/compiler/UmpleParserTest.java:452]: Where is this file?

File: cruise.umple/test/cruise/umple/compiler/008_internalConst.ump [hunk 3]
```diff
@@ -1,3 +1,3 @@
 class Group{
-    internal const String Name = " ";
+    private static final String Name = " ";
```
Comment (TimLethbridge) [inline cruise.umple/test/cruise/umple/compiler/008_internalConst.ump:2]: What does this change have to do with this PR?

File: cruise.umple/test/cruise/umple/implementation/java/ClassTemplateTest_InternalConstant.java.txt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+
+// line 1 "test.ump"
+public class Student
+{ 
+  //------------------------
+  // STATIC VARIABLES
+  //------------------------
+ 
+private static final String Name = "XYZ";
```
Comment (TimLethbridge) [inline cruise.umple/test/cruise/umple/implementation/java/ClassTemplateTest_InternalConstant.java.txt:13]: All this seems to be testing is the use of extra code

File: cruise.umple/test/cruise/umple/implementation/ClassTemplateTest_InternalConstant.ump [hunk 5]
```diff
@@ -0,0 +1,3 @@
+class Student{
+    private static final String Name = "XYZ";
```
Comment (TimLethbridge) [inline cruise.umple/test/cruise/umple/implementation/ClassTemplateTest_InternalConstant.ump:2]: I thought you had changed these to internal const. what is this file all about?

**Review by TimLethbridge — CHANGES_REQUESTED** [review]
This PR has many problems, and needs almost complete replacement

File: cruise.umple/src/Umple_Code.ump (hunk 1 above)
Comment (baala108) [reply cruise.umple/src/Umple_Code.ump:1885]: I had considered passing the counter as an argument, but while debugging noticed that the method calls: notifySubclassesAncestorImmutable, propagateImmutabilityToAllRelationships and setAncestorIsImmutable, occur in a loop. This means that the counter will reset itself and will never reach the base case.

File: cruise.umple/test/cruise/umple/compiler/UmpleParserTest.java (hunk 2 above)
Comment (baala108) [reply cruise.umple/test/cruise/umple/compiler/UmpleParserTest.java:452]: This file is under the test/cruise/umple/compiler. The content of it are the second example posted under the issue.

File: cruise.umple/test/cruise/umple/implementation/ClassTemplateTest_InternalConstant.ump (hunk 5 above)
Comment (baala108) [reply cruise.umple/test/cruise/umple/implementation/ClassTemplateTest_InternalConstant.ump:2]: This was a mistake, my editor has auto-indented the code and I had not realised that the file was staged for commit. The file is for code generation test for the internal feature I had previously worked on

File: cruise.umple/test/cruise/umple/implementation/java/ClassTemplateTest_InternalConstant.java.txt (hunk 4 above)
Comment (baala108) [reply cruise.umple/test/cruise/umple/implementation/java/ClassTemplateTest_InternalConstant.java.txt:13]: Same mistake as described above

**Pull request closed without merging**